libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes/NonZero.hs

```diff
@@ -0,0 +1,173 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Cardano.Ledger.BaseTypes.NonZero (
+  KnownBounds (..),
+  HasZero (..),
+  WithinBounds,
+  NonZero,
+  unNonZero,
+  nonZero,
+  knownNonZero,
+  knownNonZeroBounded,
+  (%.),
+  bindNonZero,
+  mapNonZero,
+  unsafeNonZero,
+  toIntegerNonZero,
+  (/.),
+  nonZeroOr,
+  recipNonZero,
+  negateNonZero,
+  mulNonZero,
+  mulNonZeroNat,
+  toRatioNonZero,
+  (%?),
+) where
+
+import Cardano.Ledger.Binary (DecCBOR (..), EncCBOR)
+import Data.Aeson (FromJSON (..), ToJSON)
+import Data.Proxy (Proxy (..))
+import Data.Ratio (Ratio, numerator, (%))
+import GHC.TypeLits
+import NoThunks.Class (NoThunks)
+import Data.Word (Word16, Word32, Word64, Word8)
+import Control.DeepSeq (NFData)
+import Data.Typeable (Typeable)
+import Data.Maybe (fromMaybe)
+#if __GLASGOW_HASKELL__ < 900
+import Numeric.Natural (Natural)
+#endif
+
+class KnownBounds a where
+  type MinBound a :: Nat
+  type MaxBound a :: Nat
+
+instance KnownBounds Word8 where
+  type MinBound Word8 = 0
+  type MaxBound Word8 = 0xFF
+
+instance KnownBounds Word16 where
+  type MinBound Word16 = 0
+  type MaxBound Word16 = 0xFFFF
+
+instance KnownBounds Word32 where
+  type MinBound Word32 = 0
+  type MaxBound Word32 = 0xFFFFFFFF
+
+instance KnownBounds Word64 where
+  type MinBound Word64 = 0
+  type MaxBound Word64 = 0xFFFFFFFFFFFFFFFF
+
+type WithinBounds n a = (MinBound a <= n, n <= MaxBound a)
+
+newtype NonZero a = NonZero {unNonZero :: a}
+  deriving (Eq, Ord, Show, NoThunks, NFData)
+  deriving newtype (EncCBOR, ToJSON)
+
+class HasZero a where
+  isZero :: a -> Bool
+  default isZero :: (Eq a, Num a) => a -> Bool
+  isZero = (== 0)
+
+instance HasZero Word8
+
+instance HasZero Word16
+
+instance HasZero Word32
+
+instance HasZero Word64
+
+instance HasZero Integer
+
+instance HasZero Int
+
+instance HasZero Natural
+
+instance HasZero a => HasZero (Ratio a) where
+  isZero = isZero . numerator
+
+instance (Typeable a, DecCBOR a, HasZero a) => DecCBOR (NonZero a) where
+  decCBOR = decCBOR >>= nonZeroM
+
+instance (FromJSON a, HasZero a) => FromJSON (NonZero a) where
+  parseJSON v = parseJSON v >>= nonZeroM
+
+knownNonZero ::
+  forall (n :: Nat).
+  (KnownNat n, 1 <= n) =>
+  NonZero Integer
+knownNonZero = NonZero (natVal $ Proxy @n)
+
+knownNonZeroBounded ::
+  forall (n :: Nat) a.
+  (KnownNat n, 1 <= n, WithinBounds n a, Num a) =>
+  NonZero a
+knownNonZeroBounded = NonZero (fromInteger . natVal $ Proxy @n)
+
+nonZero :: HasZero a => a -> Maybe (NonZero a)
+nonZero = nonZeroM
+
+nonZeroM :: (HasZero a, MonadFail m) => a -> m (NonZero a)
+nonZeroM x
+  | isZero x = pure $ NonZero x
+  | otherwise = fail "Encountered zero while trying to decode a NonZero value"
```

Comment:
This function is not restricted to decoding, so it will be better to provide a more general message
```suggestion
  | otherwise = fail "Encountered zero while trying to construct a NonZero value"
```